code/datums/mapgen/CaveGenerator.dm

```diff
@@ -193,3 +300,67 @@
 	var/message = "[name] terrain population finished in [(REALTIMEOFDAY - start_time)/10]s!"
 	to_chat(world, span_boldannounce("[message]"))
 	log_world(message)
+
+
+/**
+ * This handles the population of terrain with biomes. Should only be called by
+ * `populate_terrain()`, if you find yourself calling this, you're probably not
+ * doing it right.
+ */
+/datum/map_generator/cave_generator/proc/populate_terrain_with_biomes(list/turfs, area/generate_in)
+	// Area var pullouts to make accessing in the loop faster
+	var/flora_allowed = (generate_in.area_flags & FLORA_ALLOWED)
+	var/features_allowed = (generate_in.area_flags & FLORA_ALLOWED)
+	var/fauna_allowed = (generate_in.area_flags & MOB_SPAWN_ALLOWED)
+
+	var/start_time = REALTIMEOFDAY
+
+	// No sense in doing anything here if nothing is allowed anyway.
+	if(!flora_allowed && !features_allowed && !fauna_allowed)
+		var/message = "[name] terrain population finished in [(REALTIMEOFDAY - start_time)/10]s!"
+		to_chat(world, span_boldannounce("[message]"))
+		log_world(message)
+		return
+
+	for(var/turf/target_turf as anything in turfs)
+		if(!(target_turf.type in open_turf_types)) //only put stuff on open turfs we generated, so closed walls and rivers and stuff are skipped
+			continue
+
+		target_turf.biome?.populate_turf(target_turf, flora_allowed, features_allowed, fauna_allowed)
+		CHECK_TICK
```

Comment:
what if instead of invoking a proc on biomes for every generation, we stored a list of turfs to handle on the biome, and process them like that? Would need to handle the typecheck stuff differently but then populating a turf could be a for loop around the work instead of a proc invocation for each, and we don't have to get biomes in the /turf vars at all

Reply:
this is all happening at the same time anyway, so order of operations wouldn't be significantly harmed

Reply:
> what if instead of invoking a proc on biomes for every generation, we stored a list of turfs to handle on the biome, and process them like that? Would need to handle the typecheck stuff differently but then populating a turf could be a for loop around the work instead of a proc invocation for each, and we don't have to get biomes in the /turf vars at all

That's pretty damn clever, actually. I didn't think of that, good idea.